dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java

```diff
@@ -386,6 +381,69 @@ public File downloadTargetEntry(String toolpath, ToolDescriptor.TypeEnum type, b
         }
     }
 
+    /**
+     * Returns a list of all the tool descriptors associated with this workflow
+     * @param type workflow type, CWL, WDL, NFL ...
+     * @param entryPath Workflow path
+     * @param versionID version we are getting descriptors for
+     */
+    public List<ToolFile> getAllToolDescriptors(String type, String entryPath, String versionID) {
+
+        final Ga4Ghv20Api ga4ghv20api = this.client.getGa4Ghv20Api();
+
+        // workflows are identified within ga4gh endpoints when their prefix is '#workflow/' (tools do not require a special prefix)
+        final String ga4ghv20Path = "#workflow/" + entryPath;
+
+        // get all the tool files and filter out anything not a descriptor
+        try {
+            return ga4ghv20api.toolsIdVersionsVersionIdTypeFilesGet(type, ga4ghv20Path, versionID).stream()
+                .filter(toolFile -> toolFile.getFileType().equals(ToolFile.FileTypeEnum.SECONDARY_DESCRIPTOR) || toolFile.getFileType().equals(ToolFile.FileTypeEnum.PRIMARY_DESCRIPTOR))
+                .collect(Collectors.toList());
+        } catch (NullPointerException ex) {
+            exceptionMessage(ex, "Unable to locate entry  " + entryPath + ":" + versionID, Client.COMMAND_ERROR);
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the version ID of the given workflow, falls back to the latest version
+     * @param entryPath Workflow path
+     */
+    @Override
+    public String getVersionID(String entryPath) {
+        final String[] parts = entryPath.split(":");
+
+        final String versionID = parts.length > 1 ? parts[1] : "master";
+
+        final Workflow workflow = getDockstoreWorkflowByPath(parts[0]);
+
+        // ensure workflow has version
+        Optional<WorkflowVersion> first = workflow.getWorkflowVersions().stream().filter(foo -> foo.getName().equalsIgnoreCase(versionID))
+            .findFirst();
+
+        // if no master is present (for example, for hosted workflows), fail over to the latest descriptor
+        if (first.isEmpty()) {
+            first = workflow.getWorkflowVersions().stream().max(Comparator.comparing(WorkflowVersion::getLastModified));
+            first.ifPresent(workflowVersion -> System.out.println(
+                "Could not locate workflow with version '" + versionID + "'. Using last modified version '" + workflowVersion.getName()
+                    + "' instead."));
+            return first.get().getName();
```

Comment:
Will NPE if `first` is empty. And I think we do have edge cases of workflows with no versions.

Reply:
[Fixed](https://github.com/dockstore/dockstore-cli/blob/7f6c85a4a186819af1abeaf2f795939c1fb693b4/dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java#L435).